lib/Pico.php

```diff
@@ -0,0 +1,943 @@
+     * @return string the base url
+     */
+    public function getBaseUrl()
+    {
+        if (!empty($this->getConfig('base_url'))) {
+            return $this->getConfig('base_url');
+        }
+
+        if (
+            (!empty($_SERVER['HTTPS']) && $_SERVER['HTTPS'] != 'off')
+            || ($_SERVER['SERVER_PORT'] == 443)
+            || (!empty($_SERVER['HTTP_X_FORWARDED_PROTO']) && $_SERVER['HTTP_X_FORWARDED_PROTO'] == 'https')
+        ) {
+            $protocol = 'https';
+        } else {
+            $protocol = 'http';
+        }
+
+        $this->config['base_url'] =
+            $protocol . "://" . $_SERVER['HTTP_HOST']
+            . dirname($_SERVER['SCRIPT_NAME']) . '/';
+
+        return $this->getConfig('base_url');
+    }
+
+    /**
+     * Returns true if URL rewriting is enabled
+     *
+     * @return boolean true if URL rewriting is enabled, false otherwise
+     */
+    public function isUrlRewritingEnabled()
+    {
+        if (($this->getConfig('rewrite_url') === null) && isset($_SERVER['PICO_URL_REWRITING'])) {
+            return (bool) $_SERVER['PICO_URL_REWRITING'];
+        } elseif ($this->getConfig('rewrite_url')) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Returns the URL to a given page
+     *
+     * @param  string $page identifier of the page to link to
+     * @return string       URL
+     */
+    public function getPageUrl($page)
+    {
+        return $this->getBaseUrl() . ((!$this->isUrlRewritingEnabled() && !empty($page)) ? '?' : '') . $page;
+    }
+
+    /**
+     * Recursively walks through a directory and returns all containing files
+     * matching the specified file extension in alphabetical order
+     *
+     * @param  string $directory start directory
+     * @param  string $ext       return files with this file extension only (optional)
+     * @return array             list of found files
+     */
+    protected function getFiles($directory, $fileExtension = '')
+    {
+        $directory = rtrim($directory, '/');
+        $result = array();
+
+        // scandir() reads files in alphabetical order
+        $files = scandir($directory);
+        $fileExtensionLength = strlen($fileExtension);
+        if ($files !== false) {
+            foreach ($files as $file) {
+                // exclude hidden files/dirs starting with a .; this also excludes the special dirs . and ..
+                // exclude files ending with a ~ (vim/nano backup) or # (emacs backup)
+                if ((substr($file, 0, 1) === '.') || in_array(substr($file, -1), array('~', '#'))) {
+                    continue;
+                }
+
+                if (is_dir($directory . '/' . $file)) {
+                    // get files recursively
+                    $result = array_merge($result, $this->getFiles($directory . '/' . $file, $fileExtension));
+                } elseif (empty($fileExtension) || (substr($file, -strlen($fileExtension)) === $fileExtension)) {
```

Comment:
Should use the $fileExtensionLength variable defined in line 901 instead of the strlen() call.